### PR TITLE
[SPARK-50559][SQL] Store Except, Intersect and Union's outputs as lazy vals

### PR DESCRIPTION
**Pull request by vladimirg-db**
### What changes were proposed in this pull request?

Store `Except`, `Intersect` and `Union`'s outputs as lazy vals.

### Why are the changes needed?

Currently `Union`'s (same is for `Except` and `Intersect`) `output` is a `def`. This creates performance issues for queries with large number of stacked `UNION`s because of rules like `WidenSetOperationTypes` that traverse the logical plan and call `output` on each `Union` node. This has quadratic complexity: O(number_of_unions * (1 + 2 + 3 + ... + number_of_unions)).

Profile:
image
image

[flamegraph.tar.gz](https://github.com/user-attachments/files/18118260/flamegraph.tar.gz)

The improvement in parsing + analysis wall-clock time for a query with 500 UNIONs over 30 columns each is 13x (5.5s -> 400ms):
image

Repro:
```
def genValues(num: Int) = s"VALUES (${(0 until num).mkString(", ")})"
def genUnions(numUnions: Int, numValues: Int) = (0 until numUnions).map(_ => genValues(numValues)).mkString(" UNION ALL ")
spark.time { spark.sql(s"SELECT * FROM ${genUnions(numUnions = 500, numValues = 30)}").queryExecution.analyzed }
```

 For `EXCEPT` the perf difference is not that noticeable. Perhaps because it reuses the same `Seq` (it just calls `left.output`).

### Does this PR introduce _any_ user-facing change?

No, this is an optimization.

### How was this patch tested?

- Ran the async-profiler
- Ran the benchmark in spark-shell.
- Existing tests.

### Was this patch authored or co-authored using generative AI tooling?

copilot.nvim.

### CODE REVIEW

**Review by MaxGekk — COMMENTED** [review]
> Ran the benchmark in spark-shell.

@vladimirg-db How about to add such benchmark at `sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -376,10 +376,13 @@ case class Intersect(
 
   final override val nodePatterns: Seq[TreePattern] = Seq(INTERSECT)
 
-  override def output: Seq[Attribute] =
-    left.output.zip(right.output).map { case (leftAttr, rightAttr) =>
-      leftAttr.withNullability(leftAttr.nullable && rightAttr.nullable)
+  override def output: Seq[Attribute] = {
+    if (conf.getConf(SQLConf.INTERSECT_OPEREATOR_OUTPUT_AS_LAZY_VAL)) {
```
[the remark this answers is not in the document]
Comment (vladimirg-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:380]: I don't think there are any major risks (except increased memory usage for the `Union`/`Except`/`Intersect` nodes, but we used to allocate nodes anyway). However, this is a change to the fundamental Catalyst operators, so it has to be guarded.

**Comment by vladimirg-db** [comment]
Added the benchmarks.

**Review by MaxGekk — COMMENTED** [review]
@vladimirg-db Could you generate results for JDK21 (see other benchmark results) otherwise lgtm.

**Comment by vladimirg-db** [comment]
Added the benchmark results for JDK21

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 2]
```diff
@@ -5306,6 +5306,40 @@ object SQLConf {
...
+      .doc("When set to true, Intersect operator's output will be a lazy val. This is a " +
+        "performance for querires with a large number of stacked INTERSECTs. This is because of " +
+        "rules like WidenSetOperationTypes that traverse the logical plan tree and call output " +
+        "on each Intersect node. Such traversal has quadratic complexity: " +
+        "O(number_of_nodes * (1 + 2 + 3  + ... + number_of_nodes)).")
+      .version("4.0.0")
+      .booleanConf
+      .createWithDefault(true)
+
+  val UNION_OPEREATOR_OUTPUT_AS_LAZY_VAL = buildConf("spark.sql.unionOperatorOutputAsLazyVal")
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:5332]: Don't think need too fine granularity, how about to introduce at least one config for all operators:
```suggestion
  val LAZY_SET_OPERATOR_OUTPUT = buildConf("spark.sql.lazySetOperatorOutput.enabled")
```

Comment (vladimirg-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala]: Done.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @vladimirg-db.

**Pull request closed without merging**